File: report_generator/jira_client.py
import os
import requests
from requests.auth import HTTPBasicAuth
from datetime import date, datetime, timedelta, timezone
from dotenv import load_dotenv

load_dotenv()


class JiraClient:
    def __init__(self):
        self.base_url = os.environ["JIRA_BASE_URL"]
        self.auth = HTTPBasicAuth(os.environ["JIRA_EMAIL"], os.environ["JIRA_API_TOKEN"])
        self.board_id = os.environ["JIRA_BOARD_ID"]
        self.project_key = os.environ["JIRA_PROJECT_KEY"]
        self.velocity_sprints = int(os.getenv("VELOCITY_SPRINTS", "5"))
        self.session = requests.Session()
        self.session.auth = self.auth
        self.session.headers.update({"Accept": "application/json"})

    def _get(self, path, params=None):
        url = f"{self.base_url}{path}"
        resp = self.session.get(url, params=params, timeout=30)
        resp.raise_for_status()
        return resp.json()

    def _search(self, jql, fields, max_results=200):
        all_issues = []
        start_at = 0
        while True:
            data = self._get("/rest/api/2/search", params={
                "jql": jql,
                "fields": ",".join(fields),
                "maxResults": max_results,
                "startAt": start_at,
            })
            issues = data.get("issues", [])
            all_issues.extend(issues)
            if start_at + len(issues) >= data.get("total", 0):
                break
            start_at += len(issues)
        return all_issues

    def get_active_sprint(self):
        data = self._get(
            f"/rest/agile/1.0/board/{self.board_id}/sprint",
            params={"state": "active"},
        )
        sprints = data.get("values", [])
        if not sprints:
            raise RuntimeError(f"No active sprint found for board {self.board_id}")
        return sprints[0]

    def get_sprint_issues(self):
        jql = f"project = {self.project_key} AND sprint in openSprints()"
        fields = [
            "summary", "status", "assignee", "customfield_10016",
            "priority", "issuetype", "updated", "duedate", "flagged",
        ]
        return self._search(jql, fields)

    def get_completed_issues(self, report_date=None):
        report_day = _parse_report_date(report_date)
        period_start = datetime.combine(report_day - timedelta(days=6), datetime.min.time(), tzinfo=timezone.utc)
        period_end = datetime.combine(report_day + timedelta(days=1), datetime.min.time(), tzinfo=timezone.utc)
        jql = (
            f"project = {self.project_key} AND sprint in openSprints() "
            f'AND status = Done AND updated >= "{period_start:%Y-%m-%d %H:%M}" '
            f'AND updated < "{period_end:%Y-%m-%d %H:%M}"'
        )
        fields = ["summary", "status", "assignee", "customfield_10016", "resolutiondate", "updated"]
        return self._search(jql, fields)

    def get_blockers(self):
        jql = (
            f"project = {self.project_key} AND sprint in openSprints() "
            "AND (flagged = Impediment OR priority = Blocker)"
        )
        fields = ["summary", "assignee", "priority", "flagged", "created", "comment"]
        return self._search(jql, fields)

    def get_next_sprint_issues(self):
        jql = f"project = {self.project_key} AND sprint in futureSprints() ORDER BY priority ASC"
        fields = ["summary", "priority", "customfield_10016", "assignee"]
        return self._search(jql, fields, max_results=20)

    def get_closed_sprints(self, limit=None):
        n = limit or self.velocity_sprints
        data = self._get(
            f"/rest/agile/1.0/board/{self.board_id}/sprint",
            params={"state": "closed", "maxResults": n},
        )
        sprints = data.get("values", [])
        return sorted(sprints, key=lambda s: s.get("startDate", ""), reverse=True)[:n]

    def get_sprint_velocity(self, sprint_id):
        """Return (committed_sp, completed_sp) for a given sprint."""
        jql = f"project = {self.project_key} AND sprint = {sprint_id}"
        issues = self._search(jql, ["status", "customfield_10016"])
        committed = sum(_sp(i) for i in issues)
        completed = sum(
            _sp(i) for i in issues
            if i["fields"]["status"]["statusCategory"]["key"] == "done"
        )
        return committed, completed

    def get_bug_trend(self):
        """Return per-sprint bug open/close counts for the last N closed sprints."""
        sprints = self.get_closed_sprints()
        trend = []
        for sprint in reversed(sprints):
            sid = sprint["id"]
            opened = len(self._search(
                f"project = {self.project_key} AND sprint = {sid} AND issuetype = Bug",
                ["summary"],
            ))
            closed = len(self._search(
                f"project = {self.project_key} AND sprint = {sid} AND issuetype = Bug AND status = Done",
                ["summary"],
            ))
            trend.append({
                "sprint": sprint["name"],
                "opened": opened,
                "closed": closed,
                "net": opened - closed,
            })
        return trend


def _sp(issue):
    return issue["fields"].get("customfield_10016") or 0


def _parse_report_date(report_date):
    if report_date is None:
        return datetime.now(timezone.utc).date()
    if isinstance(report_date, date):
        return report_date
    try:
        return date.fromisoformat(report_date)
    except (TypeError, ValueError) as exc:
        raise ValueError("report_date must be an ISO YYYY-MM-DD date") from exc

File: backend/src/clients/jiraClient.js
'use strict';

const https = require('https');
const http = require('http');

function request(options, body) {
    const mod = options.protocol === 'https:' ? https : http;
    return new Promise((resolve, reject) => {
        const req = mod.request(options, (res) => {
            let data = '';
            res.on('data', chunk => { data += chunk; });
            res.on('end', () => {
                if (res.statusCode >= 400) {
                    reject(new Error(`Jira HTTP ${res.statusCode}: ${data.slice(0, 200)}`));
                } else {
                    try { resolve(JSON.parse(data)); } catch (e) { reject(new Error(`Jira parse error: ${e.message}`)); }
                }
            });
        });
        req.setTimeout(15000, () => { req.destroy(); reject(new Error('Jira request timeout')); });
        req.on('error', reject);
        if (body) req.write(body);
        req.end();
    });
}

class JiraClient {
    _init() {
        if (this._auth) return;
        const base = process.env.JIRA_BASE_URL;
        const email = process.env.JIRA_EMAIL;
        const token = process.env.JIRA_API_TOKEN;
        if (!base || !email || !token) throw new Error('Missing Jira environment variables');
        this._base = new URL(base);
        this._auth = `Basic ${Buffer.from(`${email}:${token}`).toString('base64')}`;
        this._boardId = process.env.JIRA_BOARD_ID;
        this._projectKey = process.env.JIRA_PROJECT_KEY;
    }

    _opts(path) {
        this._init();
        return {
            protocol: this._base.protocol,
            hostname: this._base.hostname,
            port: this._base.port || (this._base.protocol === 'https:' ? 443 : 80),
            path,
            method: 'GET',
            headers: { Authorization: this._auth, Accept: 'application/json' },
        };
    }

    async getActiveSprint() {
        const data = await request(this._opts(`/rest/agile/1.0/board/${this._boardId}/sprint?state=active&maxResults=1`));
        return data.values?.[0] || null;
    }

    async getCompletedSprints(count = 5) {
        const data = await request(this._opts(`/rest/agile/1.0/board/${this._boardId}/sprint?state=closed&maxResults=${count}`));
        const vals = data.values || [];
        return vals.slice(-count);
    }

    async searchIssues(jql, fields) {
        const f = fields || 'summary,status,assignee,priority,issuetype,duedate,resolutiondate,customfield_10016,customfield_10014,updated';
        const path = `/rest/api/2/search?jql=${encodeURIComponent(jql)}&maxResults=100&fields=${f}`;
        const data = await request(this._opts(path));
        return data.issues || [];
    }

    getCompletedIssues(periodStart, periodEnd) {
        this._init();
        const jql = `project = ${this._projectKey} AND sprint in openSprints() AND status = Done AND updated >= "${periodStart}" AND updated < "${periodEnd}"`;
        return this.searchIssues(jql);
    }

    getInProgressIssues() {
        this._init();
        return this.searchIssues(`project = ${this._projectKey} AND sprint in openSprints() AND status != Done`);
    }

    getBlockers() {
        this._init();
        return this.searchIssues(`project = ${this._projectKey} AND sprint in openSprints() AND (flagged = Impediment OR priority = Blocker)`);
    }

    getNextSprintIssues() {
        this._init();
        return this.searchIssues(
            `project = ${this._projectKey} AND sprint in futureSprints() ORDER BY priority ASC`,
            'summary,priority,customfield_10016,issuetype'
        );
    }
}

module.exports = new JiraClient();
